sources/LibProtection.TestSite/Default.aspx.cs

```diff
@@ -0,0 +1,188 @@
+                    new Example {FormatFunc = FormatHelper<JavaScript>, Prefix="javascript", Formatter="alert({0})", Replacer="\"/>Hello", TagBuilder = GetFormatTagBuilder("<label onclick={0}>ON CLICK</label>") },
+                    new Example {FormatFunc = FormatHelper<Sql>, Prefix="sql", Formatter="select * from myTable where myColumn = '{0}'", Replacer="' OR 1 = 1 --", TagBuilder = SqlRequestTagBuilder },
+                    new Example {FormatFunc = FormatHelper<Url>, Prefix="url", Formatter="./img/{0}", Replacer="../spanch.gif", TagBuilder = GetFormatTagBuilder("<img src=\"{0}\" />") },
+                    new Example {FormatFunc = FormatHelper<FilePath>, Prefix="filepath", Formatter=@".\files\{0}", Replacer=@"..\textFile.txt", TagBuilder = PathTagBuilder },
+                    //new Example {FormatFunc = FormatHelper<Xpath>, Prefix="xpath", Formatter=@"descendant::bk:book[bk:author='{0}']", Replacer="' or ''='", TagBuilder = XPathTagBuilder },
+                    //new Example {FormatFunc = FormatHelper<Xml>, Prefix="xml", Formatter="<?xml version=\"1.0\"?>{0}", Replacer="<!DOCTYPE foo [<!ELEMENT foo ANY><!ENTITY xxe SYSTEM \"file:///textfile.txt\">]><foo>&xxe;</foo>", TagBuilder = XMLTagBuilder },
+        };
+
+        #region Builders
+        protected static string FormatHelper<T>(string format, string[] parameters) where T : LanguageProvider
+        {
+            return SafeString<T>.TryFormat(format, out var result, parameters) ? result : null;
+        }
+
+        protected static string XPathTagBuilder(string path)
+        {
+            if (string.IsNullOrWhiteSpace(path)) { return string.Empty; }
+
+            var doc = new System.Xml.XmlDocument();
+            doc.Load(System.IO.Path.Combine(HttpRuntime.AppDomainAppPath, "xmldocument.xml"));
+            var root = doc.DocumentElement;
+
+            // Add the namespace.
+            var nsmgr = new System.Xml.XmlNamespaceManager(doc.NameTable);
+            nsmgr.AddNamespace("bk", "urn:newbooks-schema");
+
+            var nodes = root.SelectNodes(path, nsmgr);
+
+            var builder = new System.Text.StringBuilder();
+            foreach (System.Xml.XmlNode row in nodes)
+            {
+
+                builder.AppendFormat("<br>{0} - {1}", row["title"].InnerText, row["author"].InnerText);
+            }
+            return builder.ToString();
+        }
+
+        protected static string XMLTagBuilder(string xml)
+        {
+            if (string.IsNullOrWhiteSpace(xml)) { return string.Empty; }
+
+            var settings = new System.Xml.XmlReaderSettings();
+            settings.DtdProcessing = System.Xml.DtdProcessing.Parse;
+            var xmlDoc = new System.Xml.XmlDocument();
+
+            var xmlBytes = Encoding.ASCII.GetBytes(xml);
+
+            var doc = new System.Xml.XmlDocument();
```

Comment:
Duplicate instantiation (already instantiated in xmlDoc before).

Reply:
fixed